spotify_tensorflow/luigi/beam_base_task.py

```diff
@@ -0,0 +1,146 @@
+import sys
+from typing import List  # noqa: F401
+
+import luigi
+from luigi.contrib.gcs import GCSTarget, GCSFlagTarget
+
+logger = logging.getLogger("luigi-interface")
+
+
+class BeamBaseTask(luigi.Task):
+    """Base task for Beam jobs"""
+    runner = luigi.Parameter(default="DataflowRunner", description="Beam runner")
+    project = luigi.Parameter(description="GCP project for Dataflow Beam job")
+    machineWorkerType = luigi.Parameter(default="n1-standard-4", description="Dataflow worker type")
+    region = luigi.Parameter(default="europe-west1", description="Dataflow region")
+    temp_location = luigi.Parameter(description="Temporary location")
+
+    def extra_cmd_line_args(self):  # type: () -> List[str]
+        """
+        Additional command line arguments specific to subtask.
+        Override to provide.
+        """
+        return []
+
+    def run(self):
+        cmd_line = self._make_cmd_line()
+        logger.info(" ".join(cmd_line))
+
+        import subprocess
+        try:
+            return_code = self._run_with_logging(cmd_line)
+        except subprocess.CalledProcessError as e:
+            logging.error(e, exc_info=True)
+            return_code = e.returncode
+        sys.exit(return_code)
+
+    def beam_executable(self):
+        """
+        Defines the executable used to run the beam job.
+        Override to provide.
+        :return:
+        """
+        pass
+
+    def _beam_cmd_line_args(self):
+        cmd_line = [self.beam_executable()]
+
+        cmd_line.append("--runner=%s" % self.runner)
+        cmd_line.append("--project=%s" % self.project)
+        cmd_line.append("--temp_location=%s" % self.temp_location)
+        cmd_line.append("--machineWorkerType=%s" % self.machineWorkerType)
+        cmd_line.append("--region=%s" % self.region)
+
+        return cmd_line
+
+    def _make_cmd_line(self):
+        cmd = self._beam_cmd_line_args()
+        cmd.extend(self.extra_cmd_line_args())
+        return cmd
+
+    def _get_input_args(self):
+        job_input = self.input()
+        if isinstance(job_input, luigi.Target):
+            job_input = {"input": job_input}
+        if len(job_input) == 0:  # default requires()
+            return []
+        if not isinstance(job_input, dict):
+            raise ValueError("Input (requires()) must be dict type")
+        input_args = []
+        for (name, targets) in job_input.items():
+            uris = [self._get_uri(target) for target in luigi.task.flatten(targets)]
+            if isinstance(targets, dict):
+                # If targets is a dict that means it had multiple outputs. In this case make the
+                # input args "<input key>-<task output key>"
+                names = ["%s-%s" % (name, key) for key in targets.keys()]
+            else:
+                names = [name] * len(uris)
+            for (arg_name, uri) in zip(names, uris):
+                input_args.append("--%s=%s" % (arg_name, uri))
+
+        return input_args
```

Comment:
@daikeshi I'm still thinking that this is maybe not worth it. The code is quite difficult to read and it's not obvious to me as a user what I should be expecting.